controllers/mysql_clustering.go

```diff
@@ -0,0 +1,664 @@
+	rows, err := db.Unsafe().Queryx(`show master status`)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var status MySQLPrimaryStatus
+	if rows.Next() {
+		err = rows.StructScan(&status)
+		if err != nil {
+			return nil, err
+		}
+		return &status, nil
+	}
+
+	return nil, errors.New("primary status is empty")
+}
+
+func (r *MySQLClusterReconciler) getMySQLReplicaStatus(ctx context.Context, log logr.Logger, db *sqlx.DB) (*MySQLReplicaStatus, error) {
+	rows, err := db.Unsafe().Queryx(`show slave status`)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var status MySQLReplicaStatus
+	if rows.Next() {
+		err = rows.StructScan(&status)
+		if err != nil {
+			return nil, err
+		}
+		return &status, nil
+	}
+
+	return nil, nil
+}
+
+func (r *MySQLClusterReconciler) getMySQLGlobalVariablesStatus(ctx context.Context, log logr.Logger, db *sqlx.DB) (*MySQLGlobalVariablesStatus, error) {
+	rows, err := db.Queryx(`select @@read_only, @@super_read_only, @@rpl_semi_sync_master_wait_for_slave_count`)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var status MySQLGlobalVariablesStatus
+	if rows.Next() {
+		err = rows.StructScan(&status)
+		if err != nil {
+			return nil, err
+		}
+		return &status, nil
+	}
+
+	return nil, errors.New("readOnly status is empty")
+}
+
+func (r *MySQLClusterReconciler) getMySQLCloneStateStatus(ctx context.Context, log logr.Logger, db *sqlx.DB) (*MySQLCloneStateStatus, error) {
+	rows, err := db.Queryx(` select state from performance_schema.clone_status`)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var status MySQLCloneStateStatus
+	if rows.Next() {
+		err = rows.StructScan(&status)
+		if err != nil {
+			return nil, err
+		}
+		return &status, nil
+	}
+
+	return nil, nil
+}
+
+func validateConstraints(status *MySQLClusterStatus, cluster *mocov1alpha1.MySQLCluster) error {
+	if status == nil {
+		panic("unreachable condition")
+	}
+
+	var writableInstanceCounts int
+	var primaryIndex int
+	for i, status := range status.InstanceStatus {
+		if !status.GlobalVariableStatus.ReadOnly {
+			writableInstanceCounts++
+			primaryIndex = i
+		}
+	}
+	if writableInstanceCounts > 1 {
+		return moco.ErrConstraintsViolation
+	}
+
+	if cluster.Status.CurrentPrimaryIndex != nil && writableInstanceCounts == 1 {
+		if *cluster.Status.CurrentPrimaryIndex != primaryIndex {
+			return moco.ErrConstraintsViolation
+		}
+	}
+
+	cond := findCondition(cluster.Status.Conditions, mocov1alpha1.ConditionViolation)
+	if cond != nil {
+		return moco.ErrConstraintsRecovered
+	}
+
+	return nil
+}
+
+// TODO: Implementation for failover
+func selectPrimary(status *MySQLClusterStatus, cluster *mocov1alpha1.MySQLCluster) int {
+	return 0
+}
+
+func updatePrimary(cluster *mocov1alpha1.MySQLCluster, newPrimaryIndex int) []Operator {
+	currentPrimaryIndex := cluster.Status.CurrentPrimaryIndex
+	if currentPrimaryIndex != nil && *currentPrimaryIndex == newPrimaryIndex {
```

Comment:
updatePrimary doesn't change `rpl_semi_sync_master_wait_for_slave_count` when the replicas field is changed from 3 to 5.

Reply:
We will do it in the EPIC Enrich functions.